code/modules/clothing/modular/modular.dm

```diff
@@ -0,0 +1,210 @@
+	var/allowed_armor_path = list(
+		/obj/item/armor_module,
+	)
+ 	///List of allowed types [/obj/item/armor_module] that can be installed into this modular armor.
+	var/allowed_modules_path = list(
+		/obj/item/armor_module,
+	)
+	///List of allowed types [/obj/item/armor_module] that can be installed into this modular armor.
+	var/allowed_storage_path = list(
+		/obj/item/armor_module,
+	)
+
+	/// Misc stats
+	light_strength = 4
+
+
+/obj/item/clothing/suit/modular/Initialize()
+	. = ..()
+	installed_modules = list()
+
+
+/obj/item/clothing/suit/modular/Destroy()
+	QDEL_NULL(slot_chest)
+	QDEL_NULL(slot_arms)
+	QDEL_NULL(slot_legs)
+
+	QDEL_LIST(installed_modules)
+	installed_modules = null
+	QDEL_NULL(installed_storage)
+	QDEL_NULL(storage)
+	return ..()
+
+
+/obj/item/clothing/suit/modular/attack_self(mob/user)
+	if(!isturf(user.loc))
+		to_chat(user, "<span class='warning'>You cannot turn the light on while in [user.loc].</span>")
+		return
+	if(cooldowns[COOLDOWN_ARMOR_LIGHT] || !ishuman(user))
+		return
+	var/mob/living/carbon/human/H = user
+	if(H.wear_suit != src)
+		return
+	toggle_armor_light(user)
+	return TRUE
+
+/obj/item/clothing/suit/modular/item_action_slot_check(mob/user, slot)
+	if(!light_strength) // No light no ability
+		return FALSE
+	if(!ishuman(user))
+		return FALSE
+	if(slot != SLOT_WEAR_SUIT)
+		return FALSE
+	return TRUE //only give action button when armor is worn.
+
+
+/obj/item/clothing/suit/modular/attack_hand(mob/living/user)
+	if(!storage)
+		return ..()
+	if(storage.handle_attack_hand(user))
+		return ..()
+
+
+/obj/item/clothing/suit/modular/MouseDrop(over_object, src_location, over_location)
+	if(!storage)
+		return ..()
+	if(storage?.handle_mousedrop(usr, over_object))
+		return ..()
+
+
+/obj/item/clothing/suit/modular/attackby(obj/item/I, mob/user, params)
+	. = ..()
+	if(.)
+		return
+	if(!storage)
+		return
+	if(I.tool_behaviour)
+		return
+
+	return storage.attackby(I, user, params)
+
+
+/obj/item/clothing/suit/modular/emp_act(severity)
+	storage?.emp_act(severity)
+	return ..()
+
+
+/obj/item/clothing/suit/modular/screwdriver_act(mob/living/user, obj/item/I)
+	. = ..()
+	if(!length(installed_modules))
+		to_chat(user, "<span class='notice'>There is nothing to remove</span>")
+		return
+
+	var/obj/item/armor_module/attachable/attachment
+	if(length(installed_modules) == 1) // Single item (just take it)
+		attachment = installed_modules[1]
+	else if(length(installed_modules) > 1) // Multi item, ask which piece
+		attachment = input(user, "Which module would you like to remove", "Remove module") as null|anything in installed_modules
+	if(!attachment)
+		return
+
+	if(!attachment.can_detach(user, src))
+		return
+
+	attachment.on_detach(user, src)
+	update_overlays()
+
+
+/obj/item/clothing/suit/modular/crowbar_act(mob/living/user, obj/item/I)
+	. = ..()
+	var/list/obj/item/armor_module/armor/armor_slots = list(slot_chest, slot_arms, slot_legs)
+	listclearnulls(armor_slots)
```

Comment:
Do we expect null entries to be here?

Reply:
yeah, you might not have a slot_chest installed when you try to crowbar

Reply:
The how about this?
```cpp
var/list/obj/item/armor_module/armor/armor_slots = list()
if(slot_chest)
	armor_slots +=slot_chest
if(slot_arms)
	armor_slots +=slot_arms
if(slot_legs)
	armor_slots +=slot_legs
```